src/common/utxobased/engine/makeUtxoEngine.ts

```diff
@@ -344,9 +350,83 @@ export async function makeUtxoEngine(config: EngineConfig): Promise<EdgeCurrency
       return transaction
     },
 
-    sweepPrivateKeys(_spendInfo: EdgeSpendInfo): Promise<EdgeTransaction> {
-      // @ts-ignore
-      return Promise.resolve(undefined)
+    async sweepPrivateKeys(spendInfo: EdgeSpendInfo): Promise<EdgeTransaction> {
+      const privateKeys = spendInfo.privateKeys ?? []
+      if (privateKeys.length < 1) throw new Error('No private keys given')
+      let success: (
+        value: EdgeTransaction | PromiseLike<EdgeTransaction>
+      ) => void
+      let failure: (reason?: any) => void
+      const end: Promise<EdgeTransaction> = new Promise((resolve, reject) => {
+        success = resolve
+        failure = reject
+      })
+      const tmpDisklet = walletLocalDisklet
+      const tmpMetadata = await fetchMetadata(tmpDisklet)
+
+      const tmpEmitter = new EventEmitter() as any
+      tmpEmitter.on(
+        EmitterEvent.BALANCE_CHANGED,
+        async (currencyCode: string, nativeBalance: string) => {
+          metadata.balance = nativeBalance
+          await setMetadata(tmpDisklet, tmpMetadata)
+        }
+      )
+      tmpEmitter.on(
+        EmitterEvent.BLOCK_HEIGHT_CHANGED,
+        async (height: number) => {
+          metadata.lastSeenBlockHeight = height
+          await setMetadata(tmpDisklet, tmpMetadata)
+        }
+      )
+
+      const tmpProcessor = await makeProcessor({
+        disklet: tmpDisklet,
+        emitter: tmpEmitter
+      })
+      const blockBook = makeBlockBook({ emitter: tmpEmitter })
+      const tmpWalletTools = makeUtxoWalletTools({
+        keys: { wifKeys: privateKeys },
+        coin: currencyInfo.network,
+        network
+      })
+
+      tmpEmitter.on(EmitterEvent.ADDRESSES_CHECKED, async (ratio: number) => {
+        if (ratio === 1) {
+          await engineState.stop()
+          await blockBook.disconnect()
+          config.options.emitter = emitter
+          config.currencyInfo.gapLimit = currencyInfo.gapLimit
+
+          const utxos = await processor.fetchAllUtxos()
+          if (utxos === null || utxos.length < 1) {
+            throw new Error('Private key has no funds')
+          }
+          const publicAddress = this.getFreshAddress({})
+          const nativeAmount = tmpMetadata.balance
+          const options: TxOptions = { utxos, subtractFee: true }
+          spendInfo.spendTargets = [
+            { publicAddress: publicAddress.publicAddress, nativeAmount }
+          ]
+
+          this.makeSpend(spendInfo, options)
+            .then(tx => success(tx))
+            .catch(e => failure(e))
+        }
+      })
+
+      config.options.emitter = tmpEmitter
+      // hack to not overflow the wallet tools private key array
+      config.currencyInfo.gapLimit = privateKeys.length + 1
```

Comment:
This might cause problems later on.
I saw you set it back when on 'ADDRESSES_CHECKED' when the ratio is 1 but we can't really know how long it takes, or that it would actually get there.